stan/math/prim/scal/functor/reduce_sum.hpp

```diff
@@ -19,49 +19,32 @@ namespace math {
 
 namespace internal {
 
-// base definition => compile error
-template <class ReduceFunction, class M, class T, class Arg1, class Arg2,
-          class Arg3, class Arg4, class T_return_type>
+template <typename ReduceFunction, typename Enable, typename ReturnType, typename M, typename... Args>
 struct reduce_sum_impl {};
 
-template <class ReduceFunction, class M, class T, class Arg1, class Arg2,
-          class Arg3, class Arg4>
-struct reduce_sum_impl<ReduceFunction, M, T, Arg1, Arg2, Arg3, Arg4, double> {
-  using vmapped_t = std::vector<M>;
-  using arg1_t = std::vector<Arg1>;
-  using arg2_t = std::vector<Arg2>;
-  using arg3_t = std::vector<Arg3>;
-  using arg4_t = std::vector<Arg4>;
 
+template <typename ReduceFunction, typename ReturnType, typename M, typename... Args>
+struct reduce_sum_impl<ReduceFunction, require_arithmetic_t<ReturnType>, ReturnType, M, Args...> {
   struct recursive_reducer {
+    using vmapped_t = std::vector<M>;
+    std::tuple<const Args&...> args_tuple_;
```

Comment:
So that I get what's going on, if any of these args fall out of scope (like if we passed back this class to lazily evaluate later) these references would get destroyed and we'd have dangling refs?

Reply:
Yeah, instances of this struct shouldn't live beyond the lifetime of the call to parallel_sum that called it or there would be problems.